<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4.1 Analyzing viral protein sequences | DLL 2021, R section</title>
  <meta name="description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="4.1 Analyzing viral protein sequences | DLL 2021, R section" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  <meta name="github-repo" content="darachm/dll-r" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4.1 Analyzing viral protein sequences | DLL 2021, R section" />
  
  <meta name="twitter:description" content="A living syllabus for the R section of DLL 2021, at Stanford" />
  

<meta name="author" content="mesako" />
<meta name="author" content="Margaret" />
<meta name="author" content="Samson" />
<meta name="author" content="Zac" />
<meta name="author" content="darachm" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"/>
<link rel="next" href="getting-help-troubleshooting.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">DLL 2021 R workshops</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Workshop Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="workshop-goals.html"><a href="workshop-goals.html"><i class="fa fa-check"></i><b>1.1</b> Workshop goals</a></li>
<li class="chapter" data-level="1.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html"><i class="fa fa-check"></i><b>1.2</b> Structure and resources</a><ul>
<li class="chapter" data-level="1.2.1" data-path="structure-and-resources.html"><a href="structure-and-resources.html#workshop-schedule"><i class="fa fa-check"></i><b>1.2.1</b> Workshop schedule</a></li>
<li class="chapter" data-level="1.2.2" data-path="structure-and-resources.html"><a href="structure-and-resources.html#this-syllabus-and-videos"><i class="fa fa-check"></i><b>1.2.2</b> This syllabus and videos</a></li>
<li class="chapter" data-level="1.2.3" data-path="structure-and-resources.html"><a href="structure-and-resources.html#slack-channel"><i class="fa fa-check"></i><b>1.2.3</b> Slack channel</a></li>
<li class="chapter" data-level="1.2.4" data-path="structure-and-resources.html"><a href="structure-and-resources.html#synchronous-sessions"><i class="fa fa-check"></i><b>1.2.4</b> Synchronous sessions</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="workshop-expectations.html"><a href="workshop-expectations.html"><i class="fa fa-check"></i><b>1.3</b> Workshop expectations</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="day-3-intoduction-to-r.html"><a href="day-3-intoduction-to-r.html"><i class="fa fa-check"></i><b>2</b> Day 3 - intoduction to R</a><ul>
<li class="chapter" data-level="2.1" data-path="using-rstudio.html"><a href="using-rstudio.html"><i class="fa fa-check"></i><b>2.1</b> Using RStudio</a><ul>
<li class="chapter" data-level="2.1.1" data-path="using-rstudio.html"><a href="using-rstudio.html#r-scripts-in-rstudio"><i class="fa fa-check"></i><b>2.1.1</b> R Scripts in RStudio</a></li>
<li class="chapter" data-level="2.1.2" data-path="using-rstudio.html"><a href="using-rstudio.html#console-and-terminal"><i class="fa fa-check"></i><b>2.1.2</b> Console and Terminal</a></li>
<li class="chapter" data-level="2.1.3" data-path="using-rstudio.html"><a href="using-rstudio.html#environment-and-history"><i class="fa fa-check"></i><b>2.1.3</b> Environment and History</a></li>
<li class="chapter" data-level="2.1.4" data-path="using-rstudio.html"><a href="using-rstudio.html#files-plots-and-packages"><i class="fa fa-check"></i><b>2.1.4</b> Files, Plots, and Packages</a></li>
<li class="chapter" data-level="2.1.5" data-path="using-rstudio.html"><a href="using-rstudio.html#putting-it-all-together"><i class="fa fa-check"></i><b>2.1.5</b> Putting It All Together</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="using-rmds.html"><a href="using-rmds.html"><i class="fa fa-check"></i><b>2.2</b> Using Rmds</a><ul>
<li class="chapter" data-level="2.2.1" data-path="using-rmds.html"><a href="using-rmds.html#basic-structure"><i class="fa fa-check"></i><b>2.2.1</b> Basic Structure</a></li>
<li class="chapter" data-level="2.2.2" data-path="using-rmds.html"><a href="using-rmds.html#overview-of-markdown"><i class="fa fa-check"></i><b>2.2.2</b> Overview of Markdown</a></li>
<li class="chapter" data-level="2.2.3" data-path="using-rmds.html"><a href="using-rmds.html#overview-of-how-to-write-an-r-chunk"><i class="fa fa-check"></i><b>2.2.3</b> Overview of how to write an R chunk</a></li>
<li class="chapter" data-level="2.2.4" data-path="using-rmds.html"><a href="using-rmds.html#final-thoughts"><i class="fa fa-check"></i><b>2.2.4</b> Final Thoughts</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html"><i class="fa fa-check"></i><b>2.3</b> Getting started in R</a><ul>
<li class="chapter" data-level="2.3.1" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#getting-help-in-r"><i class="fa fa-check"></i><b>2.3.1</b> Getting help in R</a></li>
<li class="chapter" data-level="2.3.2" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-variables-and-data-types"><i class="fa fa-check"></i><b>2.3.2</b> Using variables and data types</a></li>
<li class="chapter" data-level="2.3.3" data-path="getting-started-in-r.html"><a href="getting-started-in-r.html#using-functions"><i class="fa fa-check"></i><b>2.3.3</b> Using functions</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html"><i class="fa fa-check"></i><b>2.4</b> Working with vectors</a><ul>
<li class="chapter" data-level="2.4.1" data-path="working-with-vectors.html"><a href="working-with-vectors.html#building-vectors"><i class="fa fa-check"></i><b>2.4.1</b> Building vectors</a></li>
<li class="chapter" data-level="2.4.2" data-path="working-with-vectors.html"><a href="working-with-vectors.html#indexing-and-subsetting-vectors"><i class="fa fa-check"></i><b>2.4.2</b> Indexing and subsetting vectors</a></li>
<li class="chapter" data-level="2.4.3" data-path="working-with-vectors.html"><a href="working-with-vectors.html#using-logic-and-logicals"><i class="fa fa-check"></i><b>2.4.3</b> Using logic and logicals</a></li>
<li class="chapter" data-level="2.4.4" data-path="working-with-vectors.html"><a href="working-with-vectors.html#modifying-vectors"><i class="fa fa-check"></i><b>2.4.4</b> Modifying vectors</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html"><i class="fa fa-check"></i><b>2.5</b> Working with dataframes</a><ul>
<li class="chapter" data-level="2.5.1" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#indexing-and-subsetting-dataframes"><i class="fa fa-check"></i><b>2.5.1</b> Indexing and subsetting dataframes</a></li>
<li class="chapter" data-level="2.5.2" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#exploring-dataframes"><i class="fa fa-check"></i><b>2.5.2</b> Exploring dataframes</a></li>
<li class="chapter" data-level="2.5.3" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#building-and-modifying-dataframes"><i class="fa fa-check"></i><b>2.5.3</b> Building and modifying dataframes</a></li>
<li class="chapter" data-level="2.5.4" data-path="working-with-dataframes.html"><a href="working-with-dataframes.html#handling-datatypes-in-dataframes"><i class="fa fa-check"></i><b>2.5.4</b> Handling datatypes in dataframes</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="day-4-tidyverse-and-visualizations.html"><a href="day-4-tidyverse-and-visualizations.html"><i class="fa fa-check"></i><b>3</b> Day 4 - tidyverse and visualizations</a><ul>
<li class="chapter" data-level="3.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html"><i class="fa fa-check"></i><b>3.1</b> Reading and processing data</a><ul>
<li class="chapter" data-level="3.1.1" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#loading-a-file-into-r"><i class="fa fa-check"></i><b>3.1.1</b> Loading a file into R</a></li>
<li class="chapter" data-level="3.1.2" data-path="reading-and-processing-data.html"><a href="reading-and-processing-data.html#handling-issues-in-data"><i class="fa fa-check"></i><b>3.1.2</b> Handling issues in data</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="using-tidyverse.html"><a href="using-tidyverse.html"><i class="fa fa-check"></i><b>3.2</b> Using tidyverse</a><ul>
<li class="chapter" data-level="3.2.1" data-path="using-tidyverse.html"><a href="using-tidyverse.html#goals-to-be-reworked-alongside-content"><i class="fa fa-check"></i><b>3.2.1</b> Goals to be reworked alongside content</a></li>
<li class="chapter" data-level="3.2.2" data-path="using-tidyverse.html"><a href="using-tidyverse.html#welcome-to-the-tidyverse"><i class="fa fa-check"></i><b>3.2.2</b> Welcome to the tidyverse!</a></li>
<li class="chapter" data-level="3.2.3" data-path="using-tidyverse.html"><a href="using-tidyverse.html#lets-start-by-making-sure-tidyverse-is-installed"><i class="fa fa-check"></i><b>3.2.3</b> Let‚Äôs start by making sure Tidyverse is installed</a></li>
<li class="chapter" data-level="3.2.4" data-path="using-tidyverse.html"><a href="using-tidyverse.html#try-tidyverse-on-the-iris-dataset-we-loaded-in-the-previous-section"><i class="fa fa-check"></i><b>3.2.4</b> Try tidyverse on the iris dataset we loaded in the previous section</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="essential-tidyverse-functions.html"><a href="essential-tidyverse-functions.html"><i class="fa fa-check"></i><b>3.3</b> Essential tidyverse functions</a><ul>
<li class="chapter" data-level="3.3.1" data-path="essential-tidyverse-functions.html"><a href="essential-tidyverse-functions.html#filtering-dataframes"><i class="fa fa-check"></i><b>3.3.1</b> Filtering dataframes</a></li>
<li class="chapter" data-level="3.3.2" data-path="essential-tidyverse-functions.html"><a href="essential-tidyverse-functions.html#creating-new-variables"><i class="fa fa-check"></i><b>3.3.2</b> Creating new variables</a></li>
<li class="chapter" data-level="3.3.3" data-path="essential-tidyverse-functions.html"><a href="essential-tidyverse-functions.html#summarizing-data"><i class="fa fa-check"></i><b>3.3.3</b> Summarizing data</a></li>
<li class="chapter" data-level="3.3.4" data-path="essential-tidyverse-functions.html"><a href="essential-tidyverse-functions.html#updating-fields-based-on-a-condition"><i class="fa fa-check"></i><b>3.3.4</b> Updating fields based on a condition</a></li>
<li class="chapter" data-level="3.3.5" data-path="essential-tidyverse-functions.html"><a href="essential-tidyverse-functions.html#reshaping-data-long-and-wide"><i class="fa fa-check"></i><b>3.3.5</b> Reshaping data: long and wide</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="using-tidyverse-1.html"><a href="using-tidyverse-1.html"><i class="fa fa-check"></i><b>3.4</b> Using tidyverse</a></li>
<li class="chapter" data-level="3.5" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html"><i class="fa fa-check"></i><b>3.5</b> Making plots with ggplot2</a><ul>
<li class="chapter" data-level="3.5.1" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#getting-started-with-a-ggplot"><i class="fa fa-check"></i><b>3.5.1</b> Getting started with a ggplot</a></li>
<li class="chapter" data-level="3.5.2" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#changing-plot-aesthetics"><i class="fa fa-check"></i><b>3.5.2</b> Changing plot aesthetics</a></li>
<li class="chapter" data-level="3.5.3" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#exploring-simple-plots"><i class="fa fa-check"></i><b>3.5.3</b> Exploring simple plots</a></li>
<li class="chapter" data-level="3.5.4" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#visualizing-between-groups"><i class="fa fa-check"></i><b>3.5.4</b> Visualizing between groups</a></li>
<li class="chapter" data-level="3.5.5" data-path="making-plots-with-ggplot2.html"><a href="making-plots-with-ggplot2.html#generating-heatmaps"><i class="fa fa-check"></i><b>3.5.5</b> Generating heatmaps</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html"><i class="fa fa-check"></i><b>3.6</b> Making scientific figures</a><ul>
<li class="chapter" data-level="3.6.1" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#plotting-error-bars"><i class="fa fa-check"></i><b>3.6.1</b> Plotting error bars</a></li>
<li class="chapter" data-level="3.6.2" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#showing-trends-in-data"><i class="fa fa-check"></i><b>3.6.2</b> Showing trends in data</a></li>
<li class="chapter" data-level="3.6.3" data-path="making-scientific-figures.html"><a href="making-scientific-figures.html#saving-figures-locally"><i class="fa fa-check"></i><b>3.6.3</b> Saving figures locally</a></li>
</ul></li>
<li class="chapter" data-level="3.7" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html"><i class="fa fa-check"></i><b>3.7</b> Applying basic stats</a><ul>
<li class="chapter" data-level="3.7.1" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#describing-quantitative-variables"><i class="fa fa-check"></i><b>3.7.1</b> Describing quantitative variables</a></li>
<li class="chapter" data-level="3.7.2" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#finding-differences-in-data"><i class="fa fa-check"></i><b>3.7.2</b> Finding differences in data</a></li>
<li class="chapter" data-level="3.7.3" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#identifying-correlations"><i class="fa fa-check"></i><b>3.7.3</b> Identifying correlations</a></li>
<li class="chapter" data-level="3.7.4" data-path="applying-basic-stats.html"><a href="applying-basic-stats.html#producing-linear-models"><i class="fa fa-check"></i><b>3.7.4</b> Producing linear models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"><a href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"><i class="fa fa-check"></i><b>4</b> Day 5 - building workflows, using packages, writing (re)usable code, sharing analyses</a><ul>
<li class="chapter" data-level="4.0.1" data-path="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html"><a href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html#what-is-a-workflow"><i class="fa fa-check"></i><b>4.0.1</b> What is a workflow?</a></li>
<li class="chapter" data-level="4.1" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html"><i class="fa fa-check"></i><b>4.1</b> Analyzing viral protein sequences</a><ul>
<li class="chapter" data-level="4.1.1" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#reading-in-tab-delimited-file"><i class="fa fa-check"></i><b>4.1.1</b> Reading in tab-delimited file</a></li>
<li class="chapter" data-level="4.1.2" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#loops---doing-a-similar-task-multiple-times"><i class="fa fa-check"></i><b>4.1.2</b> Loops - doing a similar task multiple times</a></li>
<li class="chapter" data-level="4.1.3" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#extending-the-workflow"><i class="fa fa-check"></i><b>4.1.3</b> Extending the workflow</a></li>
<li class="chapter" data-level="4.1.4" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#some-operations-on-strings"><i class="fa fa-check"></i><b>4.1.4</b> Some operations on strings</a></li>
<li class="chapter" data-level="4.1.5" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#back-to-the-motifs"><i class="fa fa-check"></i><b>4.1.5</b> Back to the motifs</a></li>
<li class="chapter" data-level="4.1.6" data-path="analyzing-viral-protein-sequences.html"><a href="analyzing-viral-protein-sequences.html#conclusion"><i class="fa fa-check"></i><b>4.1.6</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html"><i class="fa fa-check"></i><b>4.2</b> Getting help - troubleshooting</a><ul>
<li class="chapter" data-level="4.2.1" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#where-to-start-general-help"><i class="fa fa-check"></i><b>4.2.1</b> Where to start? General help</a></li>
<li class="chapter" data-level="4.2.2" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#debugging-specific-problems"><i class="fa fa-check"></i><b>4.2.2</b> Debugging specific problems</a></li>
<li class="chapter" data-level="4.2.3" data-path="getting-help-troubleshooting.html"><a href="getting-help-troubleshooting.html#minimal-reproducible-examples"><i class="fa fa-check"></i><b>4.2.3</b> Minimal reproducible examples</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="functions-as-modular-steps.html"><a href="functions-as-modular-steps.html"><i class="fa fa-check"></i><b>4.3</b> Functions as modular steps</a><ul>
<li class="chapter" data-level="4.3.1" data-path="functions-as-modular-steps.html"><a href="functions-as-modular-steps.html#tips-for-modular-workflows"><i class="fa fa-check"></i><b>4.3.1</b> Tips for modular workflows</a></li>
<li class="chapter" data-level="4.3.2" data-path="functions-as-modular-steps.html"><a href="functions-as-modular-steps.html#apply-is-a-popular-tool"><i class="fa fa-check"></i><b>4.3.2</b> Apply is a popular tool</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html"><i class="fa fa-check"></i><b>4.4</b> Packages - install and use</a><ul>
<li class="chapter" data-level="4.4.1" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#cran-is-the-canonical-r-package-resource"><i class="fa fa-check"></i><b>4.4.1</b> CRAN is the canonical R package resource</a></li>
<li class="chapter" data-level="4.4.2" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#using-packages-without-library-ing-them"><i class="fa fa-check"></i><b>4.4.2</b> Using packages without <code>library()</code>-ing them</a></li>
<li class="chapter" data-level="4.4.3" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#github-is-a-common-place-for-sharing-packges-in-development"><i class="fa fa-check"></i><b>4.4.3</b> Github is a common place for sharing packges in development</a></li>
<li class="chapter" data-level="4.4.4" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#bioconductor-is-a-great-place-for-bio-related-packages"><i class="fa fa-check"></i><b>4.4.4</b> Bioconductor is a great place for bio-related packages üò¨</a></li>
<li class="chapter" data-level="4.4.5" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#what-packages-do-you-have-already"><i class="fa fa-check"></i><b>4.4.5</b> What packages do you have already?</a></li>
<li class="chapter" data-level="4.4.6" data-path="packages-install-and-use.html"><a href="packages-install-and-use.html#finding-more-help-and-documentation"><i class="fa fa-check"></i><b>4.4.6</b> Finding more help and documentation</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="writing-reusable-code.html"><a href="writing-reusable-code.html"><i class="fa fa-check"></i><b>4.5</b> Writing (re)usable code</a><ul>
<li class="chapter" data-level="4.5.1" data-path="writing-reusable-code.html"><a href="writing-reusable-code.html#examples"><i class="fa fa-check"></i><b>4.5.1</b> Examples</a></li>
<li class="chapter" data-level="4.5.2" data-path="writing-reusable-code.html"><a href="writing-reusable-code.html#style-guides-can-be-inspiring"><i class="fa fa-check"></i><b>4.5.2</b> Style guides can be inspiring</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html"><i class="fa fa-check"></i><b>4.6</b> More <code>rmarkdown</code> for reproducibile sharing</a><ul>
<li class="chapter" data-level="4.6.1" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html#rmarkdown-package"><i class="fa fa-check"></i><b>4.6.1</b> rmarkdown package</a></li>
<li><a href="more-rmarkdown-for-reproducibile-sharing.html#section"></a></li>
<li class="chapter" data-level="4.6.2" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html#sharing-with-others"><i class="fa fa-check"></i><b>4.6.2</b> Sharing with others</a></li>
<li class="chapter" data-level="4.6.3" data-path="more-rmarkdown-for-reproducibile-sharing.html"><a href="more-rmarkdown-for-reproducibile-sharing.html#reproducibility-and-sharing"><i class="fa fa-check"></i><b>4.6.3</b> Reproducibility and sharing</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="conclusion-2.html"><a href="conclusion-2.html"><i class="fa fa-check"></i><b>4.7</b> Conclusion</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="appendix.html"><a href="appendix.html"><i class="fa fa-check"></i><b>5</b> Appendix</a><ul>
<li class="chapter" data-level="5.1" data-path="additional-tutorials.html"><a href="additional-tutorials.html"><i class="fa fa-check"></i><b>5.1</b> Additional tutorials</a><ul>
<li class="chapter" data-level="5.1.1" data-path="additional-tutorials.html"><a href="additional-tutorials.html#sub-sub-section"><i class="fa fa-check"></i><b>5.1.1</b> sub sub section</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="more-challenges.html"><a href="more-challenges.html"><i class="fa fa-check"></i><b>5.2</b> More challenges</a></li>
<li class="chapter" data-level="5.3" data-path="figures-ideas.html"><a href="figures-ideas.html"><i class="fa fa-check"></i><b>5.3</b> Figures ideas</a></li>
<li class="chapter" data-level="5.4" data-path="rmd-example-of-code-chunk.html"><a href="rmd-example-of-code-chunk.html"><i class="fa fa-check"></i><b>5.4</b> Rmd example of code chunk</a></li>
<li class="chapter" data-level="5.5" data-path="more-example-rmds.html"><a href="more-example-rmds.html"><i class="fa fa-check"></i><b>5.5</b> More example Rmds</a></li>
<li class="chapter" data-level="5.6" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html"><i class="fa fa-check"></i><b>5.6</b> How to edit this document</a><ul>
<li class="chapter" data-level="5.6.1" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html#setup"><i class="fa fa-check"></i><b>5.6.1</b> Setup</a></li>
<li class="chapter" data-level="5.6.2" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html#editwork-on-it"><i class="fa fa-check"></i><b>5.6.2</b> Edit/work on it</a></li>
<li class="chapter" data-level="5.6.3" data-path="how-to-edit-this-document.html"><a href="how-to-edit-this-document.html#publishing-onto-the-main-page-so-its-hosted-at-the-url"><i class="fa fa-check"></i><b>5.6.3</b> Publishing onto the main page, so it‚Äôs hosted at the URL</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="more-links.html"><a href="more-links.html"><i class="fa fa-check"></i><b>5.7</b> More links</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">DLL 2021, R section</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="analyzing-viral-protein-sequences" class="section level2">
<h2><span class="header-section-number">4.1</span> Analyzing viral protein sequences</h2>
<p>Let‚Äôs start with a project.</p>
<p>Viral structural proteins are dramatically more important to the general
public than they were 2 years ago - and under much more public attention.
Let‚Äôs take a cursory look at some of these.</p>
<p>What kind of data are we using for this?
We‚Äôll retrieve some determined protein sequences from NCBI databases.
We‚Äôve already done that work for you already. Here‚Äôs a short writeup of
what we did, amd we would recommend that you
always document where the data came from and how it was retrieved/processed:</p>
<blockquote>
<p>The NCBI Identital Protein Groups database was queried for ‚Äústructural‚Äù
with ‚ÄúDivision‚Äù restricted to ‚ÄúViruses‚Äù and ‚ÄúSource database‚Äù restricted
to ‚ÄúUniProtKB/Swiss-Prot‚Äù. 243 results were retrieved as FASTA files and
converted using bash commands into separate tab-delimited files.
<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a></p>
</blockquote>
<p>What format are these in?
These files are in the <code>data/viral_structural_proteins</code> folder,
they end in <code>.tsv</code>, are tab-delimited three fields, and look like this:</p>
<pre><code>Q8V433.1 Membrane protein       Bovine respiratory coronavirus (strain 98TXSF-110-LUN)  MSSVTTPAPVYTWTA...</code></pre>
<p>This is not a standard format, because I‚Äôd like to tell y‚Äôall about
reading tab-delimited files.
<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
<p>How do you inspect these files?
What bash tools do you use to look at these?</p>
<hr />
<div id="reading-in-tab-delimited-file" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Reading in tab-delimited file</h3>
<p>Let‚Äôs read one of them into R.</p>
<p>The records are ‚Äúdelimited‚Äù by tabs, so each field is tab-separated.
We‚Äôll need to use the <code>read.delim</code> function and specify a tab separator.</p>
<div id="review-of-looking-up-documentation" class="section level4">
<h4><span class="header-section-number">4.1.1.1</span> Review of looking up documentation</h4>
<p>You can look up the documentation for any named function or package by using
the <code>?function</code> syntax.</p>
<p>Sometimes, you need to use extra backticks to make it work, like <code>?`+`</code></p>
<p>If you don‚Äôt know what you‚Äôre looking for, you can search with <code>??</code>.
Extra backticks don‚Äôt hurt, and are necessary when you have spaces in the
query.</p>
<p>Once the documentation is open, you can search for text.
If you are using the R console, you can use <code>/</code> to open a search bar and
enter-key to search for it.</p>
<hr />
</div>
<div id="back-to-the-file-reading" class="section level4">
<h4><span class="header-section-number">4.1.1.2</span> Back to the file-reading‚Ä¶</h4>
<p>What does <code>?read.delim</code> say it does?
How do you set the delimiter/separator as a tab?
<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a></p>
<p>Is the function expecting there to be a file ‚Äúheader‚Äù?
( This would where the first line of the file has the name of each column )
Do we want to set this to <code>TRUE</code> or <code>FALSE</code> ?)</p>
<div class="incremental">
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb332-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_100.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb332-2" title="2">    <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F)</a>
<a class="sourceLine" id="cb332-3" title="3">viral_protein_data</a></code></pre></div>
<pre><code>##                           V1
## 1 Q8V433.1 Membrane protein 
##                                                       V2
## 1 Bovine respiratory coronavirus (strain 98TXSF-110-LUN)
##                                                                                                                                                                                                                                       V3
## 1 MSSVTTPAPVYTWTADEAIKFLKEWNFSLGIILLFITVILQFGYTSRSMFVYVIKMIILWLMWPLTIILTIFNCVYALNNVYLGFSIVFTIVAIIMWIVYFVNSIRLFIRTGSWWSFNPETNNLMCIDMKGRMYVRPIIEDYHTLTVTIIRGHLYMQGIKLGTGYSLSDLPAYVTVAKVSHLLTYKRGFLDKIGDTSGFAVYVKSKVGNYRLPSTQKGSGLDTALLRNNI</code></pre>
<ul>
<li>What is <code>viral_protein_data</code>? What kind of variable?</li>
<li>How do we find the column names?</li>
<li>What are two ways to access the protein sequence?</li>
<li>How do we calculate the number of characters in this protein sequence?
( Try searching with ‚Äú??<code>number of characters</code>‚Äù )</li>
</ul>
<p>What type is the <code>V3</code> column? What should it be?
How do we tell R to treat these strings as they are, and not convert to factors?</p>
</div>
<hr />

</div>
</div>
<div id="loops---doing-a-similar-task-multiple-times" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Loops - doing a similar task multiple times</h3>
<p>You will want to repeat this analysis for multiple file.
This automation is great, really useful for scaling your analyses.</p>
<p>Let‚Äôs do this for all the proteins.
The simplest way of doing this is to copy and paste it, and change the filename.</p>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb334-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_002.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb334-2" title="2">      <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F, <span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb334-3" title="3"><span class="kw">nchar</span>(viral_protein_data<span class="op">$</span>V3[[<span class="dv">1</span>]])</a></code></pre></div>
<pre><code>## [1] 177</code></pre>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb336-1" title="1">viral_protein_data &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/viral_structural_proteins/viral_proteins_003.tsv&quot;</span>,</a>
<a class="sourceLine" id="cb336-2" title="2">      <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F, <span class="dt">as.is=</span>T)</a>
<a class="sourceLine" id="cb336-3" title="3"><span class="kw">nchar</span>(viral_protein_data[<span class="dv">1</span>,<span class="st">&quot;V3&quot;</span>])</a></code></pre></div>
<pre><code>## [1] 137</code></pre>
<p>Go ahead and do this for all 242 proteins</p>
<hr />
<p>‚Ä¶ just kidding.</p>
<p>This is a lot of work, and each time we do this we can introduce errors.
If we ever have more files, we have to copy and paste more. If we ever want
to change an argument for all of them, we have to do each one.</p>
<p>Instead, we can work with a list of all the files available:</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb338-1" title="1"><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot; &quot;viral_proteins_001.tsv&quot; &quot;viral_proteins_002.tsv&quot;
## [4] &quot;viral_proteins_003.tsv&quot; &quot;viral_proteins_004.tsv&quot;</code></pre>
<p>Note that I put a <code>[1:5]</code> to limit it to the first 5.
You could also use <code>head()</code>. It‚Äôs a good idea to work
with a small subset of files while you are iterating
through development, then scale it up to the entirety.</p>
<p>Now how do you calculate the number of characters for each protein?</p>
<p>How do we do this for every file listed ‚Ä¶ ?</p>
<hr />
<div id="what-are-loops" class="section level4">
<h4><span class="header-section-number">4.1.2.1</span> What are loops?</h4>
<!-- video -->
<p>Loops are for running a ‚Äúcode block‚Äù as many times
as the ‚Äúcondition‚Äù determines.</p>
<ul>
<li><p>A ‚Äúcode block‚Äù is either one line of code, or multiple
lines of code surrounded by curly brackets - {}</p>
<pre><code>  {
      code &lt;- &quot;in a block&quot; 
      with &lt;- &quot;multiple lines&quot; 
  }</code></pre>
<p>The code just runs. Yep. It‚Äôs that simple.</p></li>
<li><p>A ‚Äúcondition‚Äù is an expression of code that
can either evaluate to either <code>TRUE</code> or <code>FALSE</code>, or
set a variable for each time running the code block.
This is often just before the code block, in <code>()</code> parentheses.</p></li>
</ul>
<p>The most common form of these kind of ‚Äúcontrol statements‚Äù is a <code>for</code> loop.
Other ‚Äúcontrol statements‚Äù or ‚Äúflow control statements‚Äù are
<code>while</code>, <code>repeat</code>, and <code>if</code>.</p>
<p>Let‚Äôs look up what they do, with <code>?`for`</code>
Note the back ticks! These are a trick in R to make anything
be interpreted as literally what you type, and not
any special characters. Like <code>?`+`</code>, or <code>?`?`</code></p>
<hr />
<p>Here‚Äôs an example loop:</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb341-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>) {</a>
<a class="sourceLine" id="cb341-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb341-3" title="3">}</a></code></pre></div>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4</code></pre>
<p>The pieces:</p>
<ul>
<li><p><code>(i in 1:4)</code> is what is being looped over - <code>1:4</code> is a vector of
1 through 4 that is created, and it is put one at a time into <code>i</code>
(a new variable). You need the parentheses.</p></li>
<li><p><code>{</code> and <code>}</code> denote the opening and closing brackets, specify the
‚Äúcode block‚Äù that is run each time.</p></li>
<li><p>inside this ‚Äúcode block‚Äù is <code>print(i)</code> - it prints the variable <code>i</code>,
which is set to a value of 1, 2, 3, or 4 for each loop</p></li>
</ul>
<hr />
<p>How do we loop through and print each file name?</p>
<p>What are the (1) code block and (2) loop condition ?</p>
<div class="incremental">
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb343-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb343-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb343-3" title="3">}</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot;
## [1] &quot;viral_proteins_001.tsv&quot;
## [1] &quot;viral_proteins_002.tsv&quot;
## [1] &quot;viral_proteins_003.tsv&quot;
## [1] &quot;viral_proteins_004.tsv&quot;</code></pre>
<p>How do we modify this to calculate the protein length?
I add a line where we use <code>nchar</code>, but <code>i</code> is the filename instead of me
typing it in there.</p>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb345-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb345-2" title="2">  <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb345-3" title="3">  <span class="kw">print</span>(<span class="kw">nchar</span>(<span class="kw">read.delim</span>(i)<span class="op">$</span>V3[[<span class="dv">1</span>]]))</a>
<a class="sourceLine" id="cb345-4" title="4">}</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot;</code></pre>
<pre><code>## Warning in file(file, &quot;rt&quot;): cannot open file &#39;viral_proteins_000.tsv&#39;: No such
## file or directory</code></pre>
<pre><code>## Error in file(file, &quot;rt&quot;): cannot open the connection</code></pre>
</div>
<p>Error! It is looking for a file <code>viral_proteins_000.tsv</code>, but it is looking
in <em>this directory</em>. It is actually in <code>data/viral_structural_proteins</code>.
Look up the <code>list.files</code> documentation, and find how to get it to return
the full name of the file.</p>
<hr />
<p>Next, this should work‚Ä¶</p>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb349-1" title="1"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb349-2" title="2">                     <span class="dt">full.names=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>] ) {</a>
<a class="sourceLine" id="cb349-3" title="3">  <span class="kw">print</span>(<span class="kw">nchar</span>(<span class="kw">read.delim</span>(i,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">header=</span>F,<span class="dt">as.is=</span>T)<span class="op">$</span>V3[[<span class="dv">1</span>]]))</a>
<a class="sourceLine" id="cb349-4" title="4">}</a></code></pre></div>
<pre><code>## [1] 1248
## [1] 1254
## [1] 177
## [1] 137
## [1] 117</code></pre>
<p>And yep, we have lengths of protein sequences.</p>
<p>‚Ä¶</p>
<p>What now?</p>
<hr />
</div>
<div id="storing-values-from-a-loop" class="section level4">
<h4><span class="header-section-number">4.1.2.2</span> Storing values from a loop</h4>
<p>How do we store these values for later analysis?</p>
<p>In other languages, ‚Äúappend‚Äù. But R is not built that way. It‚Äôll work,
but it‚Äôs very inefficient. The ‚ÄúR-way‚Äù to store values from a loop
is to define a vector of the right length,
then put each element in it.</p>
<p>Here‚Äôs some example vectors that we can create.</p>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb351-1" title="1"><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="dv">10</span>)</a></code></pre></div>
<pre><code>##  [1] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot;</code></pre>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb353-1" title="1"><span class="kw">vector</span>(<span class="dt">mode=</span><span class="st">&quot;numeric&quot;</span>,<span class="dt">length=</span><span class="dv">5</span>)</a></code></pre></div>
<pre><code>## [1] 0 0 0 0 0</code></pre>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb355-1" title="1"><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="dv">2</span>)</a></code></pre></div>
<pre><code>## [1] FALSE FALSE</code></pre>
<hr />
<p>Okay, but how do we access each position to save the value?
We need to turn out list of files into indicies.
We‚Äôll save it first so we can count how many there are.</p>
<p><code>seq_along</code> is handy function to create a number sequence along a vector,
otherwise use something like <code>seq(1,length(x))</code>.</p>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb357-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb357-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb357-3" title="3"></a>
<a class="sourceLine" id="cb357-4" title="4"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb357-5" title="5"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb357-6" title="6">    <span class="kw">print</span>(i)</a>
<a class="sourceLine" id="cb357-7" title="7">    <span class="kw">print</span>(first_five[i])</a>
<a class="sourceLine" id="cb357-8" title="8">}</a></code></pre></div>
<pre><code>## [1] 1
## [1] &quot;viral_proteins_000.tsv&quot;
## [1] 2
## [1] &quot;viral_proteins_001.tsv&quot;
## [1] 3
## [1] &quot;viral_proteins_002.tsv&quot;
## [1] 4
## [1] &quot;viral_proteins_003.tsv&quot;
## [1] 5
## [1] &quot;viral_proteins_004.tsv&quot;</code></pre>
<hr />
<p>Putting these together, we can create and save a vector of file names:</p>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb359-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb359-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb359-3" title="3"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb359-4" title="4">filenamez &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb359-5" title="5"></a>
<a class="sourceLine" id="cb359-6" title="6"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb359-7" title="7"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb359-8" title="8">  filenamez[i] &lt;-<span class="st"> </span>first_five[i]</a>
<a class="sourceLine" id="cb359-9" title="9">}</a>
<a class="sourceLine" id="cb359-10" title="10"></a>
<a class="sourceLine" id="cb359-11" title="11">filenamez</a></code></pre></div>
<pre><code>## [1] &quot;viral_proteins_000.tsv&quot; &quot;viral_proteins_001.tsv&quot; &quot;viral_proteins_002.tsv&quot;
## [4] &quot;viral_proteins_003.tsv&quot; &quot;viral_proteins_004.tsv&quot;</code></pre>
<hr />
<p>However‚Ä¶ there‚Äôs some other files in there, isn‚Äôt there?
A <code>viral_structural_proteins.fasta</code>?</p>
<p>Let‚Äôs ignore that from the <code>list.files</code> call by using the argument
<code>pattern=".*.tsv"</code> argument.
And we‚Äôll get back to the FASTA later‚Ä¶</p>
<hr />
<p>And finally calculate the length of each protein:</p>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb361-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb361-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb361-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a>
<a class="sourceLine" id="cb361-4" title="4"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb361-5" title="5">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb361-6" title="6"></a>
<a class="sourceLine" id="cb361-7" title="7"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb361-8" title="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb361-9" title="9">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb361-10" title="10">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb361-11" title="11">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb361-12" title="12">    )</a>
<a class="sourceLine" id="cb361-13" title="13">}</a>
<a class="sourceLine" id="cb361-14" title="14"></a>
<a class="sourceLine" id="cb361-15" title="15">lengthz</a></code></pre></div>
<pre><code>## [1] &quot;1248&quot; &quot;1254&quot; &quot;177&quot;  &quot;137&quot;  &quot;117&quot;</code></pre>
<hr />
<p>Now we can take off the <code>[1:5]</code> limiter, and do the whole set:</p>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb363-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb363-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb363-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb363-4" title="4"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb363-5" title="5">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;character&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb363-6" title="6"></a>
<a class="sourceLine" id="cb363-7" title="7"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb363-8" title="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb363-9" title="9">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb363-10" title="10">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb363-11" title="11">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb363-12" title="12">    )</a>
<a class="sourceLine" id="cb363-13" title="13">}</a></code></pre></div>
<hr />
<p>How do we go about visualizing/analyzing this?</p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb364-1" title="1"><span class="kw">hist</span>(lengthz)</a></code></pre></div>
<pre><code>## Error in hist.default(lengthz): &#39;x&#39; must be numeric</code></pre>
<p>er what‚Ä¶? Debug! What is <code>lengthz</code>?</p>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb366-1" title="1"><span class="kw">str</span>(lengthz)</a></code></pre></div>
<pre><code>##  chr [1:244] &quot;1248&quot; &quot;1254&quot; &quot;177&quot; &quot;137&quot; &quot;117&quot; &quot;117&quot; &quot;858&quot; &quot;1242&quot; &quot;1255&quot; ...</code></pre>
<p>Character? Let‚Äôs try numeric instead‚Ä¶</p>
<hr />
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb368-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb368-2" title="2">first_five &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb368-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb368-4" title="4"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb368-5" title="5">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;numeric&quot;</span>,<span class="kw">length</span>(first_five))</a>
<a class="sourceLine" id="cb368-6" title="6"></a>
<a class="sourceLine" id="cb368-7" title="7"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb368-8" title="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(first_five) ) {</a>
<a class="sourceLine" id="cb368-9" title="9">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb368-10" title="10">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( <span class="kw">read.delim</span>(first_five[i],</a>
<a class="sourceLine" id="cb368-11" title="11">      <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb368-12" title="12">    )</a>
<a class="sourceLine" id="cb368-13" title="13">}</a></code></pre></div>
<hr />
<p>Base R histogram</p>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb369-1" title="1"><span class="kw">hist</span>(lengthz,<span class="dt">breaks=</span><span class="dv">50</span>)</a></code></pre></div>
<p><img src="515-loops_files/figure-html/base_viz_protein_lengths_fix-1.png" width="672" /></p>
<hr />
<p>More fancy ggplot</p>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb370-1" title="1"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb370-2" title="2"><span class="kw">ggplot</span>( <span class="kw">data.frame</span>(<span class="dt">length=</span>lengthz) )<span class="op">+</span><span class="kw">theme_classic</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb370-3" title="3"><span class="st">  </span><span class="kw">aes</span>(<span class="dt">x=</span>length)<span class="op">+</span><span class="kw">geom_histogram</span>(<span class="dt">bins=</span><span class="dv">50</span>)</a></code></pre></div>
<p><img src="515-loops_files/figure-html/ggplot_viz_protein_lengths-1.png" width="672" /></p>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb371-1" title="1"><span class="kw">summary</span>(lengthz)</a></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##     0.0   230.0   626.0   726.3  1088.5  3390.0</code></pre>
<hr />
<p>And that‚Äôs about how long the viral structural proteins in this database
tend to be.</p>
<ul>
<li>How do we extend this analysis to add more files?</li>
<li>How do we calculate new properties about the sequences?</li>
<li>How do we store that?</li>
</ul>

</div>
</div>
<div id="extending-the-workflow" class="section level3">
<h3><span class="header-section-number">4.1.3</span> Extending the workflow</h3>
<p>How do you think about/plan the workflow we‚Äôve built?</p>
<p>One way is to flatten out all the tasks,
and to script each individual task every time it is done.
This requires the author, user, and reader to understand a lot of complexity.</p>
<div class="incremental">
<div class="mermaid">
graph TD
    classDef default line-height:12px;
    classDef three line-height:12px,fill:yellow;
    you[analyst] --> B[finding files] &    C[making a vector] &    D[looping] &    E[reading files] &    F[getting protein seq] &    G[calculting and saving nchar] &    H[plotting nchar]
    B:::three;
    C:::three;
    D:::three;
    E:::three;
    F:::three;
    G:::three;
    H:::three;
</div>
<p>Another way to approach this is to cluster them into a hierarchy of modules.</p>
<div class="mermaid">
graph TD
    classDef default line-height:12px;
    classDef one  line-height:12px,fill:pink;
    classDef two line-height:12px,fill:cyan;
    classDef three line-height:12px,fill:yellow;
    you2[analyst] --> find & read & plot;
    find:::one --> B;
    B[finding files]:::three;
    read:::one --> looping & reading;
    looping:::two --> C & D;
    C[making a vector]:::three;
    D[looping]:::three;
    reading[reading and calc]:::two --> E & F & G;
    E[reading files]:::three;
    F[getting protein seq]:::three;
    G[calculting and saving nchar]:::three;
    plot:::one --> H;
    H[plotting nchar]:::three;
</div>
<p>In this organization, the analyst can operate at levels of
<span style="color:magenta">steps</span>,
<span style="color:cyan">modules</span>, and
<span style="color:gold">specific instructions</span>,
depending on what is needed.</p>
</div>
<hr />
<p>Organizing your workflows into composable modules lets you extend these to
un-ancipiated complexity. You could imaging using these steps or modules:</p>
<div class="mermaid">
graph LR
    classDef default line-height:12px;
    vp[read viral proteins] --> cl[calculate lengths] --> hist[histogram]
</div>
<p>in new ways by composing the elements together, to analyze a different source
of proteins, with a new analysis, with similar plots:</p>
<div class="mermaid">
graph LR
    classDef default line-height:12px;
    vp[read viral proteins] --> cl[calculate lengths] --> hist[histogram]
    hp[read human proteins] --> cl --> boxplot[boxplot]
    vp --> cmotif[find a particular motif] --> motifhist[histogram]
    hp --> cmotif --> motifboxplot[boxplot]
</div>
<p>Where maybe these are how the inputs/outputs are defined:</p>
<pre><code>reading_proteins:
    input: directory path of proteins to read
    output: protein sequences
calculating lengths:
    input: protein sequences
    output: numeric vector, of lengths
finding a motif:
    input: protein sequences
    output: numeric vector, of presence/absence</code></pre>
<hr />
<div id="looking-for-a-smoking-motif" class="section level4">
<h4><span class="header-section-number">4.1.3.1</span> Looking for a smoking motif‚Ä¶</h4>
<p>Some folks believe the insertion of an amino-acid sequence of <code>PRRA</code>, relative
to other related coronaviruses, is very rare and a clear sign of engineering.
I don‚Äôt know hardly anything about furin-cleavage sites or viral evolution,
but we can use the tools we have to look for this sequence in the files.</p>
<p>Let‚Äôs do that<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a>:</p>
<ol style="list-style-type: decimal">
<li><p>Figure out how to search for ‚ÄúPRRA‚Äù in a character string of a protein
sequence.</p>
<ul>
<li>try using <code>grepl</code>, like so: <code>grepl("PRRA",protein_sequence)</code>
where <code>protein_sequence</code> is the variable containing the
protein sequence (similar to what you put into <code>nchar</code> before!).</li>
</ul></li>
<li><p>Modify your <code>for</code> loop to now look for this motif instead of calculating
number of characters.
Save the values in a ‚Äúnumeric‚Äù vector.
<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a></p></li>
<li><p>Plot the results. How often do you find this motif in this dataset?</p></li>
</ol>
</div>
</div>
<div id="some-operations-on-strings" class="section level3">
<h3><span class="header-section-number">4.1.4</span> Some operations on strings</h3>
<p>Character strings are a common type of data, and there‚Äôs lots of ways to
cut, dice, extract, and recognize elements to help your work.
Here‚Äôs some ideas:</p>
<div id="base-r-tools" class="section level4">
<h4><span class="header-section-number">4.1.4.1</span> Base R tools</h4>
<p>Let‚Äôs start with one of those above sentences.
You can make a string by putting characters in between single or double quotes:</p>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb374-1" title="1">stringz &lt;-<span class="st"> &quot;Character strings are a common type of data, and there&#39;s lots of ways to&quot;</span></a>
<a class="sourceLine" id="cb374-2" title="2">stringz</a></code></pre></div>
<pre><code>## [1] &quot;Character strings are a common type of data, and there&#39;s lots of ways to&quot;</code></pre>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb376-1" title="1"><span class="kw">is</span>(stringz)</a></code></pre></div>
<pre><code>## [1] &quot;character&quot;           &quot;vector&quot;              &quot;data.frameRowLabels&quot;
## [4] &quot;SuperClassMethod&quot;</code></pre>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb378-1" title="1"><span class="kw">str</span>(stringz)</a></code></pre></div>
<pre><code>##  chr &quot;Character strings are a common type of data, and there&#39;s lots of ways to&quot;</code></pre>
<p><code>grep</code> and <code>grepl</code> are good for searching for patterns (like <code>grep</code> in bash).
The first returns position, the second returns <code>TRUE</code> or <code>FALSE</code>.
What kind of value is that then?</p>
<p>I only ever use <code>grepl</code> anymore. It searches for a <code>pattern</code> string
in a character vector <code>x</code>.</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb380-1" title="1"><span class="kw">grepl</span>(<span class="dt">pattern=</span><span class="st">&quot;common type&quot;</span>,<span class="dt">x=</span>stringz)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb382-1" title="1"><span class="kw">grepl</span>(<span class="st">&quot;rare type&quot;</span>,stringz)</a></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb384-1" title="1"><span class="kw">grepl</span>(<span class="st">&quot; [comn]* type&quot;</span>,stringz)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>You can use complex
<a href="https://bookdown.org/rdpeng/rprogdatascience/regular-expressions.html">regular expressions</a>
to specify more complex patterns.
These can get
<a href="https://r4ds.had.co.nz/strings.html#matching-patterns-with-regular-expressions">really complex</a>
and <a href="https://cs.lmu.edu/~ray/notes/regex/">really powerful</a>.
We‚Äôre not going to explain these here, except just for the example below.</p>
<p>Ask on Slack later if you‚Äôd like more help with something specific.</p>
<p>Also of use is <code>sub</code> and <code>gsub</code>.
These substitute patterns with replacements, for the
first occurrence (<code>sub</code>) or globally (<code>gsub</code>).
These are really handy for modifying data-tables.</p>
<p>For example, let‚Äôs say you‚Äôd like to split up filenames by dates:</p>
<div class="sourceCode" id="cb386"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb386-1" title="1">datar &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">stringsAsFactors=</span>F,</a>
<a class="sourceLine" id="cb386-2" title="2">    <span class="dt">filenames=</span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb386-3" title="3">        <span class="st">&quot;210607_pilot_transformation_works.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb386-4" title="4">        <span class="st">&quot;210609_transformation_seems_to_work.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb386-5" title="5">        <span class="st">&quot;210610_failed_gel_images.jpeg&quot;</span></a>
<a class="sourceLine" id="cb386-6" title="6">    )</a>
<a class="sourceLine" id="cb386-7" title="7">)</a>
<a class="sourceLine" id="cb386-8" title="8">datar<span class="op">$</span>dates &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d)_(.*)</span><span class="ch">\\</span><span class="st">.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb386-9" title="9">        <span class="dt">replacement=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">1&quot;</span>,<span class="dt">x=</span>datar<span class="op">$</span>filenames)</a>
<a class="sourceLine" id="cb386-10" title="10">datar<span class="op">$</span>names &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern=</span><span class="st">&quot;(</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d)_(.*)</span><span class="ch">\\</span><span class="st">.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb386-11" title="11">        <span class="dt">replacement=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">2&quot;</span>,<span class="dt">x=</span>datar<span class="op">$</span>filenames)</a>
<a class="sourceLine" id="cb386-12" title="12">datar</a></code></pre></div>
<pre><code>##                                  filenames  dates                        names
## 1   210607_pilot_transformation_works.jpeg 210607   pilot_transformation_works
## 2 210609_transformation_seems_to_work.jpeg 210609 transformation_seems_to_work
## 3            210610_failed_gel_images.jpeg 210610            failed_gel_images</code></pre>
<p>The above uses regular expressions, where <code>\\d</code> is matching any digit,
<code>.*</code> matches anything, and <code>()</code> denote ‚Äúgroups‚Äù to ‚Äúcapture‚Äù.
In the ‚Äúreplacement‚Äù, these ‚Äúgroups‚Äù are referred to by <code>\\1</code> or <code>\\2</code>.</p>
<p>There are better (cleaner) ways of doing this‚Ä¶</p>
</div>
<div id="stringr-and-the-tidyverse" class="section level4">
<h4><span class="header-section-number">4.1.4.2</span> stringr and the tidyverse</h4>
<p>The <code>tidyverse</code> includes a package <code>stringr</code>. This cleans up some of these
basic string operations, makes them more standardized, and gives them nice
standardized names. It‚Äôs worth using, but you should know about the other
base R functions!</p>
<div class="sourceCode" id="cb388"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb388-1" title="1"><span class="kw">library</span>(stringr)</a></code></pre></div>
<p>For example the above,</p>
<div class="sourceCode" id="cb389"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb389-1" title="1">datar &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">stringsAsFactors=</span>F,</a>
<a class="sourceLine" id="cb389-2" title="2">    <span class="dt">filenames=</span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb389-3" title="3">        <span class="st">&quot;210607_pilot_transformation_works.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb389-4" title="4">        <span class="st">&quot;210609_transformation_seems_to_work.jpeg&quot;</span>,</a>
<a class="sourceLine" id="cb389-5" title="5">        <span class="st">&quot;210610_failed_gel_images.jpeg&quot;</span></a>
<a class="sourceLine" id="cb389-6" title="6">    )</a>
<a class="sourceLine" id="cb389-7" title="7">)</a>
<a class="sourceLine" id="cb389-8" title="8">datar<span class="op">$</span>dates &lt;-<span class="st"> </span><span class="kw">str_extract</span>(<span class="dt">string=</span>datar<span class="op">$</span>filenames,<span class="dt">pattern=</span><span class="st">&quot;</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d</span><span class="ch">\\</span><span class="st">d_&quot;</span>)</a>
<a class="sourceLine" id="cb389-9" title="9">datar</a></code></pre></div>
<pre><code>##                                  filenames   dates
## 1   210607_pilot_transformation_works.jpeg 210607_
## 2 210609_transformation_seems_to_work.jpeg 210609_
## 3            210610_failed_gel_images.jpeg 210610_</code></pre>
<p>There are many options, type <code>str_</code> and then TAB.</p>
<p>For much much more about strings, check out
<a href="https://r4ds.had.co.nz/strings.html">Wickham‚Äôs R4DS book</a>.
Especially for the special characters part</p>
<hr />
</div>
</div>
<div id="back-to-the-motifs" class="section level3">
<h3><span class="header-section-number">4.1.5</span> Back to the motifs</h3>
<p>Let‚Äôs go ahead and look for that sequence, ‚ÄúPRAR‚Äù in the structural proteins
we‚Äôve downloaded here. First, we make one example of that code.</p>
<p>We can start with making a positive and negative control so we can debug
this easily.</p>
<div class="sourceCode" id="cb391"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb391-1" title="1">test_string_positive &lt;-<span class="st"> &quot;APQPRARCGPGP&quot;</span></a>
<a class="sourceLine" id="cb391-2" title="2">test_string_negative &lt;-<span class="st"> &quot;AVASTMEMATEY&quot;</span></a>
<a class="sourceLine" id="cb391-3" title="3"></a>
<a class="sourceLine" id="cb391-4" title="4"><span class="kw">grepl</span>(<span class="dt">pattern=</span><span class="st">&quot;PRAR&quot;</span>,<span class="dt">x=</span>test_string_positive)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb393"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb393-1" title="1"><span class="kw">grepl</span>(<span class="dt">pattern=</span><span class="st">&quot;PRAR&quot;</span>,<span class="dt">x=</span>test_string_negative)</a></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<p>Looks good.</p>
<hr />
<p>Now let‚Äôs modify our previous code to use this line, and store it into
the appropriate position.</p>
<p>How do we do this?</p>
<div class="sourceCode" id="cb395"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb395-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb395-2" title="2">all_files &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb395-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb395-4" title="4"></a>
<a class="sourceLine" id="cb395-5" title="5"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb395-6" title="6">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;numeric&quot;</span>,<span class="kw">length</span>(all_files))</a>
<a class="sourceLine" id="cb395-7" title="7">motifz  &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="kw">length</span>(all_files))</a>
<a class="sourceLine" id="cb395-8" title="8"></a>
<a class="sourceLine" id="cb395-9" title="9"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb395-10" title="10"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(all_files) ) {</a>
<a class="sourceLine" id="cb395-11" title="11">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb395-12" title="12">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( </a>
<a class="sourceLine" id="cb395-13" title="13">        <span class="kw">read.delim</span>(all_files[i],</a>
<a class="sourceLine" id="cb395-14" title="14">            <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb395-15" title="15">    )</a>
<a class="sourceLine" id="cb395-16" title="16">  <span class="co"># looking for motif and storing it</span></a>
<a class="sourceLine" id="cb395-17" title="17">  motifz[i]  &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">pattern=</span><span class="st">&quot;PRAR&quot;</span>,</a>
<a class="sourceLine" id="cb395-18" title="18">        <span class="dt">x=</span><span class="kw">read.delim</span>(all_files[i],</a>
<a class="sourceLine" id="cb395-19" title="19">            <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb395-20" title="20">    )</a>
<a class="sourceLine" id="cb395-21" title="21">}</a></code></pre></div>
<p>Wow! This seems like it‚Äôs a really rare motif. I wonder what this means,
this may be a smoking gun, this has implic‚Ä¶.</p>
<hr />
<p>Oh wait, that‚Äôs a typo. It‚Äôs ‚ÄúPRRA‚Äù not ‚ÄúPRAR‚Äù.</p>
<div class="sourceCode" id="cb396"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb396-1" title="1"><span class="co"># listing files</span></a>
<a class="sourceLine" id="cb396-2" title="2">all_files &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="dt">path=</span><span class="st">&quot;data/viral_structural_proteins&quot;</span>,</a>
<a class="sourceLine" id="cb396-3" title="3">                         <span class="dt">full.names=</span>T, <span class="dt">pattern=</span><span class="st">&quot;.*tsv&quot;</span>)</a>
<a class="sourceLine" id="cb396-4" title="4"></a>
<a class="sourceLine" id="cb396-5" title="5"><span class="co"># initializing vectors</span></a>
<a class="sourceLine" id="cb396-6" title="6">lengthz &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;numeric&quot;</span>,<span class="kw">length</span>(all_files))</a>
<a class="sourceLine" id="cb396-7" title="7">motifz  &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;logical&quot;</span>,<span class="kw">length</span>(all_files))</a>
<a class="sourceLine" id="cb396-8" title="8"></a>
<a class="sourceLine" id="cb396-9" title="9"><span class="co"># looping through files</span></a>
<a class="sourceLine" id="cb396-10" title="10"><span class="cf">for</span> (i <span class="cf">in</span> <span class="kw">seq_along</span>(all_files) ) {</a>
<a class="sourceLine" id="cb396-11" title="11">  <span class="co"># calculating length and storing it</span></a>
<a class="sourceLine" id="cb396-12" title="12">  lengthz[i] &lt;-<span class="st"> </span><span class="kw">nchar</span>( </a>
<a class="sourceLine" id="cb396-13" title="13">        <span class="kw">read.delim</span>(all_files[i],</a>
<a class="sourceLine" id="cb396-14" title="14">            <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb396-15" title="15">    )</a>
<a class="sourceLine" id="cb396-16" title="16">  <span class="co"># looking for motif and storing it</span></a>
<a class="sourceLine" id="cb396-17" title="17">  motifz[i]  &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="dt">pattern=</span><span class="st">&quot;PRRA&quot;</span>,</a>
<a class="sourceLine" id="cb396-18" title="18">        <span class="dt">x=</span><span class="kw">read.delim</span>(all_files[i],</a>
<a class="sourceLine" id="cb396-19" title="19">            <span class="dt">header=</span>F,<span class="dt">as.is=</span>T,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)<span class="op">$</span>V3[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb396-20" title="20">    )</a>
<a class="sourceLine" id="cb396-21" title="21">}</a></code></pre></div>
<p>It actually pops up a few times. In what files?</p>
<hr />
<p>Since <code>motifz</code> is logical, we can use that to index the files vector:</p>
<div class="sourceCode" id="cb397"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb397-1" title="1">all_files[motifz]</a></code></pre></div>
<pre><code>##  [1] &quot;data/viral_structural_proteins/viral_proteins_016.tsv&quot;
##  [2] &quot;data/viral_structural_proteins/viral_proteins_020.tsv&quot;
##  [3] &quot;data/viral_structural_proteins/viral_proteins_075.tsv&quot;
##  [4] &quot;data/viral_structural_proteins/viral_proteins_076.tsv&quot;
##  [5] &quot;data/viral_structural_proteins/viral_proteins_080.tsv&quot;
##  [6] &quot;data/viral_structural_proteins/viral_proteins_109.tsv&quot;
##  [7] &quot;data/viral_structural_proteins/viral_proteins_118.tsv&quot;
##  [8] &quot;data/viral_structural_proteins/viral_proteins_119.tsv&quot;
##  [9] &quot;data/viral_structural_proteins/viral_proteins_191.tsv&quot;
## [10] &quot;data/viral_structural_proteins/viral_proteins_196.tsv&quot;
## [11] &quot;data/viral_structural_proteins/viral_proteins_197.tsv&quot;</code></pre>
<p>Additional questions:</p>
<ul>
<li><p>In what viruses, what proteins does it pop up?
How does that affect your interpretation?</p>
<p>How do you use the skills from Day1/2 to look?</p></li>
<li>Does it show up where you‚Äôd expect it to?
What might that suggest about this dataset?</li>
<li>How might we come up with a null model for this?
How often would this occur? Under what assumptions?
How do you test that?</li>
<li><p>Your null model might depend on how often certain amino acids tend to
occur. How might you write some R code to count up each occurance
of each amino acid?</p>
<p>Here‚Äôs a complex line of code that may be adaptable to this problem:</p>
<pre><code>  table(
      factor(
          unlist(
              strsplit(&quot;ACABA&quot;,split=&quot;&quot;)
          ),
          levels=c(&quot;A&quot;,&quot;B&quot;,&quot;C&quot;,&quot;D&quot;,&quot;E&quot;,&quot;F&quot;,&quot;G&quot;)
      )
  )</code></pre>
<p><code>table</code> tabulates the occurances of a string/factor,
<code>factor</code> with a <code>levels</code> argument can create ‚Äúspaces‚Äù for unobserved
letters/strings,
<code>unlist</code> turns a list into not a list,
<code>strsplit</code> breaks up a string into a vector of strings using a pattern
that can also be "".</p>
<p>I wonder what would happen if these vectors were created for each
protein, and then summed‚Ä¶</p></li>
</ul>
</div>
<div id="conclusion" class="section level3">
<h3><span class="header-section-number">4.1.6</span> Conclusion</h3>
<ul>
<li>You can use loops to organize repetitive tasks, such as working through a
large list of files and reading them.</li>
<li>You can analyze strings using functions like <code>nchar</code> or <code>grepl</code>.</li>
<li>Modular design of workflows, using functions or just organized and
commented code, can help you extend your analysis to handle new
questions as they arise</li>
</ul>
<p><br/>
<br/>
<br/>
<br/>
<br/></p>
<hr />
<script src="https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js"></script>
<script>
mermaid.initialize({
    startOnLoad: false, // don't render initially, instead....
    logLevel: 3, // for debugging
});
function initMermaid(s) { 
  var diagrams = document.querySelectorAll('.mermaid');
  var i;
  for(i=0;i<diagrams.length;i++){
    if(diagrams[i].offsetWidth>0){
      mermaid.init(undefined, diagrams[i]);
    }
  }
}
initMermaid();
</script>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="3">
<li id="fn3"><p>One crucial aspect missing is the actual bash commands used. I did not
save them as a script, and I <em>should</em> go back, save them in a script,
and re-run that to make sure I get the same result!<a href="analyzing-viral-protein-sequences.html#fnref3" class="footnote-back">‚Ü©</a></p></li>
<li id="fn4"><p>The formats you‚Äôre likely going to have to deal with are CSV, TSV,
FASTQ, FASTA, SAM - at least. <a href="analyzing-viral-protein-sequences.html#fnref4" class="footnote-back">‚Ü©</a></p></li>
<li id="fn5"><p>Often this is typed into computers and scripts as ", as the tab key
will often not put a TAB character into where you‚Äôre typing.<a href="analyzing-viral-protein-sequences.html#fnref5" class="footnote-back">‚Ü©</a></p></li>
<li id="fn6"><p>
‚Ä¶since we‚Äôre just demonstrating R coding and not representing ourselves as
being familiar enough with the field to properly interpret these results.
I have no idea what occurance of this would mean or not mean, or are we
doing any sort of statistics or comparisons to null expectations.<a href="analyzing-viral-protein-sequences.html#fnref6" class="footnote-back">‚Ü©</a></p></li>
<li id="fn7"><p>Why not logical, since these are logical results? It makes the <code>hist()</code>
step slightly easier.<a href="analyzing-viral-protein-sequences.html#fnref7" class="footnote-back">‚Ü©</a></p></li>
</ol>
</div>
<p style="margin-top:10em;text-align:center;font-size:small;">
    <a href="https://github.com/darachm/dll-r/blob/main/LICENSE.md">
        Licensed Creative Commons Attribution-NonCommercial-ShareAlike 4.0
    </a>
</p>
<p style="text-align:center;font-size:small;">
    <a href="https://github.com/rstudio/bookdown" target="blank">
        Website typeset with bookdown
    </a>
</p>

<script> 
// This is to look through, everything that's a floating lone double dash,
// delete it. This is left over for stuff that's trying to be pauses for
// the slides
var listz = document.getElementsByTagName("p");
for (var i=0; i < listz.length; i++) { 
    if (listz[i].innerText === "‚Äì") { listz[i].innerText = "" } 
} 
</script>
            </section>

          </div>
        </div>
      </div>
<a href="day-5-building-workflows-using-packages-writing-reusable-code-sharing-analyses.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="getting-help-troubleshooting.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/darachm/dll-r/edit/main/510-viral-protein-analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["DLL 2021 R syllabus site.html"],
"toc": {
"collapse": "subsection"
},
"highlight": "zenburn"
});
});
</script>

</body>

</html>
